config/config.go

```diff
@@ -302,24 +306,24 @@ func (c Config) String() string {
 	return string(b)
 }
 
-// GetScrapeConfigs returns the scrape configurations.
+// GetScrapeConfigs returns the read-only, validated scrape configurations including
+// the ones from the scrape_config_files.
+// This method does not write to config, and it's concurrency safe (the pointer receiver is for efficiency).
+// This method also assumes the Config was created by Load, due to races,
+// read mode https://github.com/prometheus/prometheus/issues/15538.
 func (c *Config) GetScrapeConfigs() ([]*ScrapeConfig, error) {
-	scfgs := make([]*ScrapeConfig, len(c.ScrapeConfigs))
+	if !c.loaded {
+		return nil, errors.New("main config scrape configs was not validated and loaded; GetScrapeConfigs method can only be used on configuration from the config.Load or config.LoadFile")
```

Comment:
Is there any chance we can log this error to a user? Sounds very developer-focused

Reply:
There is, but should be caught by the error. What's the alternative you suggest? (:

Reply:
Would we prefer panic here? I would ok with that.

Otherwise I think we could indeed make it obvious and say "config was not validated and loaded; programmatic error, should never happened").

If we don't check here, there will be more confusing error later on (e.g. "interval should be 0" because globals were not updated). Revalidation risk races (:

Reply:
Yes, a panic seems better as this would be a misuse and it's better to fail earlier.

Reply:
One way to avoid this is to make `config.Config` private, so no one can create it and `Load` will have to be used. This will be no go idiomatic, but might work. Otherwise the best long term solution is to split config into type that's parsed and validated and then to type that's actually used in code (without global section etc).

Reply:
Improved the message and commentary.